### PR TITLE
Remove random writes from SST file ingestion

### PR DESCRIPTION
**Pull request by riversand963**
RocksDB used to store global_seqno in external SST files written by
SstFileWriter. During file ingestion, RocksDB uses `pwrite` to update the
`global_seqno`. Since random write is not supported in some non-POSIX compliant
file systems, external SST file ingestion is not supported on these file
systems. To address this limitation, we no longer update `global_seqno` during
file ingestion. Later RocksDB uses the MANIFEST and other information in table
properties to deduce global seqno for externally-ingested SST files.

Test plan:
```
$make clean && make -j16 all check
$./external_sst_file_test
$./external_sst_file_basic_test

### CODE REVIEW

**Comment by riversand963** [comment]
@siying Could you PTAL? Thanks

File: db/version_edit.h [hunk 1]
```diff
@@ -36,18 +36,26 @@ struct FileDescriptor {
...
   uint64_t file_size;  // File size in bytes
+  SequenceNumber smallest_seqno;  // The smallest seqno in this file
+  SequenceNumber largest_seqno;   // The largest seqno in this file
 
   FileDescriptor() : FileDescriptor(0, 0, 0) {}
 
-  FileDescriptor(uint64_t number, uint32_t path_id, uint64_t _file_size)
+  FileDescriptor(uint64_t number, uint32_t path_id, uint64_t _file_size,
+                 SequenceNumber _smallest_seqno = kMaxSequenceNumber,
+                 SequenceNumber _largest_seqno = 0)
```
Comment (siying) [inline db/version_edit.h]: How hard it is to avoid defaults?

File: table/block_based_table_reader.cc [hunk 2]
```diff
@@ -696,7 +707,7 @@ SequenceNumber GetGlobalSequenceNumber(const TableProperties& table_properties,
     return kDisableGlobalSequenceNumber;
   }
 
-  SequenceNumber global_seqno = DecodeFixed64(seqno_pos->second.c_str());
+  SequenceNumber global_seqno = largest_seqno;
```
Comment (siying) [inline table/block_based_table_reader.cc]: Can we still read from `DecodeFixed64(seqno_pos->second.c_str())` and make sure it is either the same as `largest_seqno` or it is not modified?

Comment (riversand963) [reply table/block_based_table_reader.cc]: by 'not modified', do you mean verify the checksum?

File: db/version_edit.h (hunk 1 above)
Comment (riversand963) [reply db/version_edit.h]: Done.

**Comment by riversand963** [comment]
@siying can you PTAL?

File: table/block_based_table_reader.cc [hunk 3]
```diff
@@ -697,6 +697,8 @@ SequenceNumber GetGlobalSequenceNumber(const TableProperties& table_properties,
   }
 
   SequenceNumber global_seqno = DecodeFixed64(seqno_pos->second.c_str());
```
Comment (siying) [inline table/block_based_table_reader.cc]: I'm surprised that the existing code has no verification that seqno_pos is valid.

File: table/table_builder.h [hunk 4]
```diff
@@ -31,14 +32,15 @@ struct TableReaderOptions {
                      const EnvOptions& _env_options,
                      const InternalKeyComparator& _internal_comparator,
                      bool _skip_filters = false, bool _immortal = false,
-                     int _level = -1)
+                     int _level = -1, SequenceNumber _largest_seqno = 0)
```
[the remark this answers is not in the document]
Comment (riversand963) [reply table/table_builder.h]: Do you mean defaults of all the parameters?

[the remark this answers is not in the document]
Comment (siying) [reply table/table_builder.h]: I don't care level, but I don't think it makes sense to have a default for largest_seqno. I don't care whether you reorder to two so you can keep the level default, or remove defaults for both of level and largest_seqno.

File: table/block_based_table_reader.cc (hunk 3 above)
Comment (riversand963) [reply table/block_based_table_reader.cc]: Done.

File: table/table_builder.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (riversand963) [reply table/table_builder.h]: Done. Adding an overloaded constructor. Another option is to provide a setter (not implemented).

File: table/block_based_table_reader.cc [hunk 5]
```diff
@@ -697,6 +697,8 @@ SequenceNumber GetGlobalSequenceNumber(const TableProperties& table_properties,
...
+  assert(global_seqno == 0 || global_seqno == largest_seqno);
```
[the remark this answers is not in the document]
Comment (riversand963) [reply table/block_based_table_reader.cc]: Done.

File: include/rocksdb/options.h [hunk 6]
```diff
@@ -1265,6 +1265,9 @@ struct IngestExternalFileOptions {
   // with allow_ingest_behind=true since the dawn of time.
   // All files will be ingested at the bottommost level with seqno=0.
   bool ingest_behind = false;
+  // Set to true if you would like to write global_seqno to a given offset in
+  // the external SST file.
```
[the remark this answers is not in the document]
Comment (riversand963) [reply include/rocksdb/options.h]: Done.

**Pull request closed without merging**